server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3767,4 +3610,242 @@ synchronized boolean assertConsistent() {
             return true;
         }
     }
+
+    private static final class CreateSnapshotTask implements ClusterStateTaskListener {
+
+        private final Repository repository;
+
+        private final RepositoryData repositoryData;
+
+        private final ActionListener<Snapshot> listener;
+
+        private final Snapshot snapshot;
+
+        private final CreateSnapshotRequest createSnapshotRequest;
+
+        private final RepositoryMetadata repositoryMetadata;
+
+        CreateSnapshotTask(
+            ActionListener<Snapshot> listener,
+            Snapshot snapshot,
+            RepositoryData repositoryData,
+            Repository repository,
+            CreateSnapshotRequest createSnapshotRequest,
+            RepositoryMetadata initialRepositoryMetadata
+        ) {
+            this.listener = listener;
+            this.snapshot = snapshot;
+            this.repositoryData = repositoryData;
+            this.repository = repository;
+            this.createSnapshotRequest = createSnapshotRequest;
+            this.repositoryMetadata = initialRepositoryMetadata;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            logger.warn(
+                () -> format("[%s][%s] failed to create snapshot", snapshot.getRepository(), snapshot.getSnapshotId().getName()),
+                e
+            );
+            listener.onFailure(e);
+        }
+    }
+
+    private class CreateSnapshotTaskExecutor implements ClusterStateTaskExecutor<CreateSnapshotTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<CreateSnapshotTask> batchExecutionContext) {
+            final ClusterState state = batchExecutionContext.initialState();
+            final SnapshotsInProgress initialSnapshots = state.custom(SnapshotsInProgress.TYPE, SnapshotsInProgress.EMPTY);
+            SnapshotsInProgress snapshotsInProgress = initialSnapshots;
+            for (final var taskContext : batchExecutionContext.taskContexts()) {
+                try {
+                    final var task = taskContext.getTask();
+                    final var repoMeta = state.metadata()
+                        .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY)
+                        .repository(task.snapshot.getRepository());
+                    if (Objects.equals(task.repositoryMetadata, repoMeta)) {
+                        snapshotsInProgress = createSnapshot(taskContext, state, snapshotsInProgress);
+                    } else {
+                        // repository data changed in between starting the task and executing this cluster state update so try again
+                        submitCreateSnapshotRequest(task.createSnapshotRequest, task.listener, task.repository, task.snapshot, repoMeta);
+                        taskContext.success(() -> {});
```

Comment:
I'd rather spawn this task at the end of the publication rather than within the executor:

```suggestion
                        taskContext.success(
                            () -> submitCreateSnapshotRequest(
                                task.createSnapshotRequest,
                                task.listener,
                                task.repository,
                                task.snapshot,
                                repoMeta
                            )
                        );
```

Reply:
The `#success()` callback still happens within this one publication so the repo metadata it captures will be the same either way.

My thinking here is that the actual execution should be a pure computation almost always - if we stick to this, we'll be able to do some nice optimisations in future (e.g. run a few future updates updates speculatively while we're waiting for the current one to commit)

Reply:
Ah fair point, we did have that idea at one point :) -> moved to the success callback now.